File: MapReduce and Spark/python/task1.py
# input_file = "test_review.json"
#input_file = "ccc_72aa618f64_45625@runweb11:/mnt/data2/students/sub1/ccc_72aa618f64_45625/asn325078_1/asn325119_1/resource/asnlib/publicdata/business.json"
from pyspark import SparkContext
import json
import sys



# def main(review_filepath, output_file_path, n_partition):
def main(review_filepath, output_file_path):
    sc = SparkContext('local[*]', 'task1')
    # textRDD = sc.textFile(review_filepath, int(n_partition))
    textRDD = sc.textFile(review_filepath)
    ## 1a

    tmp = textRDD.map(lambda line : (1,1)).reduceByKey(lambda a, b : a+b).collect()
    n_review = tmp[0][1]

    ## 1b

    def ftn1(line):
        year=int(json.loads(line)["date"][0:4])
        if year == 2018:
            return 1
        else:
            return 0
    tmp = textRDD.map(lambda line : (1,ftn1(line))).reduceByKey(lambda a, b : a+b).collect()
    n_review_2018 = tmp[0][1]

    ## 1c

    tmp = textRDD.map(lambda line : (json.loads(line)["user_id"],1)).reduceByKey(lambda a, b : 1).map(lambda x:(1,1)).reduceByKey(lambda a,b : a+b).collect()
    n_user = tmp[0][1]

    ## 1d
    tmp = textRDD.map(lambda line: (json.loads(line)["user_id"], 1)).reduceByKey(lambda a, b: a + b) \
        .map(lambda x: (x[1], x[0])).groupByKey().sortByKey(False).mapValues(lambda v: sorted(v)).flatMap(
        lambda x: [(x[0], value) for value in x[1]]).take(10)

    tmp_dict = {}
    for i in tmp:
        tmp_dict[i[1]]=i[0]

    tmp_dict = sorted(tmp_dict.items(), key=lambda x: (-x[1], x[0]))
    top10_user = []
    for i in tmp_dict:
        top10_user.append([i[0],i[1]])

    ## 1e

    tmp = textRDD.map(lambda line : (json.loads(line)["business_id"],1)).reduceByKey(lambda a,b : a)\
    .map(lambda x : (1,1)).reduceByKey(lambda a,b : a+b).collect()
    n_business = tmp[0][1]

    ## 1f

    tmp = textRDD.map(lambda line: (json.loads(line)["business_id"], 1)).reduceByKey(lambda a, b: a + b) \
        .map(lambda x: (x[1], x[0])).groupByKey().sortByKey(False).mapValues(lambda v: sorted(v)).flatMap(
        lambda x: [(x[0], value) for value in x[1]]).take(10)

    tmp_dict = {}
    for i in tmp:
        tmp_dict[i[1]]=i[0]

    tmp_dict = sorted(tmp_dict.items(), key=lambda x: (-x[1], x[0]))

    top10_business = []
    for i in tmp_dict:
        top10_business.append([i[0],i[1]])



    ## Final result

    answer_json = {}
    answer_json["n_review"] = n_review
    answer_json["n_review_2018"] = n_review_2018
    answer_json["n_user"] = n_user
    answer_json["top10_user"] = top10_user
    answer_json["n_business"] = n_business
    answer_json["top10_business"] = top10_business

    with open(output_file_path, 'w') as json_file:
        json.dump(answer_json, json_file)

review_filepath = sys.argv[1]
output_file_path = sys.argv[2]
# n_partition = sys.argv[3]

# main(review_filepath, output_file_path, n_partition)
main(review_filepath, output_file_path)

File: 1. MapReduce and Spark/scala/task1.scala
import org.apache.spark.{SparkConf, SparkContext}
import org.json4s._
import org.json4s.jackson.JsonMethods._

import java.io._
import org.json4s._
import scala.util.parsing.json.JSONObject


object task1{
  def main(args:Array[String]): Unit= {

    // get argument
    val review_filepath = args(0)
    val output_filepath = args(1)
//    val review_filepath = "/Users/jaeyoungkim/IdeaProjects/hw1/src/main/scala/test_review.json"

    val spark = new SparkConf().setAppName("task1").setMaster("local[*]")
    val sc = new SparkContext(spark)
    val textRDD = sc.textFile(path = review_filepath)


    //1a
    val tmp1 = textRDD.map(line => (1, 1)).reduceByKey(_ + _).collect()

    val n_review = tmp1(0)._2

    println(n_review)


    //1b

    def ftn1(line: String): Int = {
      val data = parse(line)
      val date = data \ "date"
      val JString(year_) = date
      val year = year_.substring(0, 4)
      if (year == "2018") {
        return 1;
      }
      else {
        return 0;
      }

    }


    val tmp2 = textRDD.map(line => (1, ftn1(line))).reduceByKey(_ + _).collect()
    val n_review_2018 = tmp2(0)._2

    println(n_review_2018)


    //1c

    def ftn2(line: String): String = {
      val data = parse(line)
      val date = data \ "user_id"
      val JString(user_id) = date
      return user_id
    }

    (1, 1)

    def ftn3(x: Int, y: Int): Int = {
      return 1
    }

    val tmp3 = textRDD.map(line => (ftn2(line), 1)).reduceByKey(ftn3).map(x => (1, 1)).reduceByKey(_ + _).collect()
    val n_user = tmp3(0)._2

    println(n_user)

    //1d


    def ftn4(line: (Int, List[String])): List[(String, Int)] = {
      val num = line._1
      val list_of_names = line._2
      var res: List[(String, Int)] = List()
      for (w <- 0 to list_of_names.size - 1) {
        //        println((num, list_of_names(w)))
        //        res :+ (num, list_of_names(w))
        res = res :+ ((list_of_names(w), num))
      }
      return res
    }

    val tmp4 = textRDD.map(line => (ftn2(line), 1)).reduceByKey(_ + _).map(x => (x._2, x._1)).groupByKey().sortByKey(false).mapValues(v => v.toList.sortWith(_ < _)).flatMap(x => ftn4(x)).take(10)
    //    println(tmp4)
    var top10_user: List[Any] = List()
    for (w <- 0 to tmp4.size - 1) {
      top10_user = top10_user :+ ("\""+tmp4(w)._1+"\"", tmp4(w)._2)
    }
    println(top10_user)


    //1e


    def ftn5(line: String): String = {
      val data = parse(line)
      val date = data \ "business_id"
      val JString(business_id) = date
      return business_id
    }

    def ftn6(x: Int, y: Int): Int = {
      return x
    }

    val tmp5 = textRDD.map(line => (ftn5(line), 1)).reduceByKey(ftn6).map(x => (1, 1)).reduceByKey(_ + _).collect()
//    println(tmp5)
    val n_business = tmp5(0)._2
//    println(n_business)


    //    1f


    val tmp6 = textRDD.map(line => (ftn5(line), 1)).reduceByKey(_ + _).map(x => (x._2, x._1)).groupByKey().sortByKey(false).mapValues(v => v.toList.sortWith(_ < _)).flatMap(x => ftn4(x)).take(10)
//    println(tmp6)
    var top10_business: List[Any] = List()
    for (w <- 0 to tmp6.size - 1) {
      top10_business = top10_business :+ ("\""+tmp6(w)._1+"\"", tmp6(w)._2)
    }

    println(top10_user)

    val answer = Map("n_review" -> n_review, "n_review_2018" -> n_review_2018, "n_user" -> n_user, "top10_user" -> top10_user, "n_business" -> n_business, "top10_business" -> top10_business)


    val answer_string = JSONObject(answer).toString()
//    println(answer)
    val answer_string_revised = answer_string.replace('(', '[').replace(')',']').replaceAll("List", "")
    // FileWriter



    val file = new File(output_filepath)
    val bw = new BufferedWriter(new FileWriter(file))
    bw.write(answer_string_revised)
    bw.close()





  }
}
